File: internal/scaffold/render_test.go
package scaffold

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"cup/internal/tmpl"
)

// writeTemplate drops a project-local override template so Render/tmpl.Read read
// deterministic content instead of a built-in.
func writeTemplate(t *testing.T, root, kind, name, content string) {
	t.Helper()
	dir := filepath.Join(root, tmpl.ProjectTemplateDir, kind, name)
	if err := os.MkdirAll(filepath.Dir(dir), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(dir, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
}

func TestRenderSubstitutes(t *testing.T) {
	root := t.TempDir()
	writeTemplate(t, root, "class", "source.h.tmpl", "class {{Name}} in {{ns}} {};")

	got, err := Render(root, "headers", "class", "source.h.tmpl", map[string]string{
		"Name": "Widget",
		"ns":   "ui",
	})
	if err != nil {
		t.Fatalf("Render returned error: %v", err)
	}
	if want := "class Widget in ui {};"; got != want {
		t.Errorf("Render = %q, want %q", got, want)
	}
}

// A variable whose value contains another placeholder must be resolved to a
// fixed point.
func TestRenderResolvesNestedPlaceholders(t *testing.T) {
	root := t.TempDir()
	writeTemplate(t, root, "greet", "hello.tmpl", "{{greeting}}")

	got, err := Render(root, "headers", "greet", "hello.tmpl", map[string]string{
		"greeting": "hi {{name}}",
		"name":     "cup",
	})
	if err != nil {
		t.Fatalf("Render returned error: %v", err)
	}
	if want := "hi cup"; got != want {
		t.Errorf("Render = %q, want %q", got, want)
	}
}

func TestRenderUnresolvedPlaceholder(t *testing.T) {
	root := t.TempDir()
	writeTemplate(t, root, "class", "source.h.tmpl", "{{Name}} {{Missing}} {{Missing}}")

	_, err := Render(root, "headers", "class", "source.h.tmpl", map[string]string{"Name": "X"})
	if err == nil {
		t.Fatal("Render = nil error, want unresolved-placeholder error")
	}
	if !strings.Contains(err.Error(), "{{Missing}}") {
		t.Errorf("error %q should name the unresolved placeholder", err)
	}
	// The unresolved set is de-duplicated, so {{Missing}} appears once.
	if strings.Count(err.Error(), "{{Missing}}") != 1 {
		t.Errorf("error %q should list {{Missing}} exactly once", err)
	}
}

func TestRenderMissingTemplate(t *testing.T) {
	root := t.TempDir()
	if _, err := Render(root, "headers", "nope", "source.h.tmpl", nil); err == nil {
		t.Fatal("Render for missing template = nil error, want error")
	}
}

func TestEnsureFile(t *testing.T) {
	root := t.TempDir()
	path := filepath.Join(root, "sub", "dir", "file.txt")

	if err := EnsureFile(root, path, "first"); err != nil {
		t.Fatalf("EnsureFile (create) returned error: %v", err)
	}
	if b, _ := os.ReadFile(path); string(b) != "first" {
		t.Errorf("file content = %q, want %q", b, "first")
	}

	// A second call must not overwrite existing content.
	if err := EnsureFile(root, path, "second"); err != nil {
		t.Fatalf("EnsureFile (existing) returned error: %v", err)
	}
	if b, _ := os.ReadFile(path); string(b) != "first" {
		t.Errorf("EnsureFile overwrote existing file: content = %q, want %q", b, "first")
	}
}

func TestWriteFileCreatesNested(t *testing.T) {
	root := t.TempDir()
	path := filepath.Join(root, "a", "b", "c.txt")

	// A fresh path never prompts, so this exercises the non-interactive path.
	wrote, err := WriteFile(root, path, "content")
	if err != nil {
		t.Fatalf("WriteFile returned error: %v", err)
	}
	if !wrote {
		t.Error("WriteFile reported not written for a new file")
	}
	if b, _ := os.ReadFile(path); string(b) != "content" {
		t.Errorf("file content = %q, want %q", b, "content")
	}
}

File: cpp/src/tests/render_test.cpp
#include <catch2/catch_test_macros.hpp>

#include <filesystem>
#include <fstream>
#include <map>
#include <string>

#include "TempDir.hpp"

import cup.scaffold;
import cup.tmpl;

namespace {

using cup::test::TempDir;

// Drops a project-local override template so render()/tmpl::read read
// deterministic content instead of a built-in.
void write_template(const TempDir& root, std::string_view kind, std::string_view name,
                     std::string_view content) {
    root.write(std::filesystem::path(cup::tmpl::kProjectTemplateDir) / kind / name, content);
}

[[nodiscard]] std::string read_file(const std::filesystem::path& path) {
    std::ifstream in(path, std::ios::binary);
    return std::string(std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>());
}

}

TEST_CASE("render substitutes every {{key}}", "[render]") {
    const TempDir root;
    write_template(root, "class", "source.h.tmpl", "class {{Name}} in {{ns}} {};");

    const auto got = cup::scaffold::render(root, "headers", "class", "source.h.tmpl",
                                           {{"Name", "Widget"}, {"ns", "ui"}});
    REQUIRE(got.has_value());
    REQUIRE(*got == "class Widget in ui {};");
}

TEST_CASE("render resolves a placeholder nested inside a variable's value", "[render]") {
    const TempDir root;
    write_template(root, "greet", "hello.tmpl", "{{greeting}}");

    const auto got = cup::scaffold::render(root, "headers", "greet", "hello.tmpl",
                                           {{"greeting", "hi {{name}}"}, {"name", "cup"}});
    REQUIRE(got.has_value());
    REQUIRE(*got == "hi cup");
}

TEST_CASE("render reports every unresolved placeholder, deduplicated", "[render]") {
    const TempDir root;
    write_template(root, "class", "source.h.tmpl", "{{Name}} {{Missing}} {{Missing}}");

    const auto got =
        cup::scaffold::render(root, "headers", "class", "source.h.tmpl", {{"Name", "X"}});
    REQUIRE_FALSE(got.has_value());
    const std::string& message = got.error().message();
    const auto first = message.find("{{Missing}}");
    REQUIRE(first != std::string::npos);
    REQUIRE(message.find("{{Missing}}", first + 1) == std::string::npos);
}

TEST_CASE("render reports a missing template", "[render]") {
    const TempDir root;
    REQUIRE_FALSE(cup::scaffold::render(root, "headers", "nope", "source.h.tmpl", {}).has_value());
}

TEST_CASE("ensure_file creates a file but never overwrites an existing one", "[render]") {
    const TempDir root;
    const std::filesystem::path path = root.path() / "sub" / "dir" / "file.txt";

    REQUIRE(cup::scaffold::ensure_file(root, path, "first").has_value());
    REQUIRE(read_file(path) == "first");

    REQUIRE(cup::scaffold::ensure_file(root, path, "second").has_value());
    REQUIRE(read_file(path) == "first");
}

TEST_CASE("write_file creates nested directories for a fresh path without prompting",
          "[render]") {
    const TempDir root;
    const std::filesystem::path path = root.path() / "a" / "b" / "c.txt";

    // A fresh path never prompts, so this exercises the non-interactive path.
    const auto wrote = cup::scaffold::write_file(root, path, "content");
    REQUIRE(wrote.has_value());
    REQUIRE(*wrote);
    REQUIRE(read_file(path) == "content");
}
